crepecake/src/main/java/com/google/cloud/tools/crepecake/builder/BuildConfiguration.java

```diff
@@ -0,0 +1,151 @@
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.builder;
+
+import com.google.cloud.tools.crepecake.registry.DockerCredentialRetriever;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Map;
+
+/** Immutable configuration options for the builder process. */
+public class BuildConfiguration {
+
+  /** Enumerates the fields in the configuration. */
+  @VisibleForTesting
+  enum Fields {
+    /** The server URL of the registry to pull the base image from. */
+    BASE_IMAGE_SERVER_URL,
+    /** The image name/repository the base image (also known as the registry namespace). */
+    BASE_IMAGE_NAME,
+
+    /** The server URL of the registry to push the built image to. */
+    TARGET_SERVER_URL,
+    /** The image name/repository of the built image (also known as the registry namespace). */
+    TARGET_IMAGE_NAME,
+    /** The image tag of the built image (the part after the colon). */
+    TARGET_TAG,
+
+    /** The credential helper name used by {@link DockerCredentialRetriever}. */
+    CREDENTIAL_HELPER_NAME,
+  }
+
+  public static class Builder {
+
+    /** Textual descriptions of the configuration fields. */
+    @VisibleForTesting
+    static final Map<Fields, String> FIELD_DESCRIPTIONS =
+        new EnumMap<Fields, String>(Fields.class) {
+          {
+            put(Fields.BASE_IMAGE_SERVER_URL, "base image registry server URL");
+            put(Fields.BASE_IMAGE_NAME, "base image name");
+            put(Fields.TARGET_SERVER_URL, "target registry server URL");
+            put(Fields.TARGET_IMAGE_NAME, "target image name");
+            put(Fields.TARGET_TAG, "target image tag");
+            put(Fields.CREDENTIAL_HELPER_NAME, "credential helper name");
+          }
+        };
+
+    private Map<Fields, String> values = new EnumMap<>(Fields.class);
+
+    private Builder() {}
+
+    public Builder setBaseImageServerUrl(String baseImageServerUrl) {
+      values.put(Fields.BASE_IMAGE_SERVER_URL, baseImageServerUrl);
+      return this;
+    }
+
+    public Builder setBaseImageName(String baseImageName) {
+      values.put(Fields.BASE_IMAGE_NAME, baseImageName);
+      return this;
+    }
+
+    public Builder setTargetServerUrl(String targetServerUrl) {
+      values.put(Fields.TARGET_SERVER_URL, targetServerUrl);
+      return this;
+    }
+
+    public Builder setTargetImageName(String targetImageName) {
+      values.put(Fields.TARGET_IMAGE_NAME, targetImageName);
+      return this;
+    }
+
+    public Builder setTargetTag(String targetTag) {
+      values.put(Fields.TARGET_TAG, targetTag);
+      return this;
+    }
+
+    public Builder setCredentialHelperName(String credentialHelperName) {
+      values.put(Fields.CREDENTIAL_HELPER_NAME, credentialHelperName);
+      return this;
+    }
+
+    public BuildConfiguration build() throws BuildConfigurationMissingValueException {
```

Comment:
Per Effective Java 2e, Item 60, you should use an IllegalStateException for this case.